### PR TITLE
Add tests for interactions with books

### PR DESCRIPTION
**Pull request by Lessorg**

### CODE REVIEW

File: src/test/java/test/project/bookshop/controller/BookControllerTests.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class BookControllerTests {
+    protected static MockMvc mockMvc;
+
+    @MockBean
+    private BookService bookService;
+
+    @MockBean
+    private JwtUtil jwtUtil;
```
Comment (liliia-ponomarenko) [inline src/test/java/test/project/bookshop/controller/BookControllerTests.java]: Why do you need these mocks?

File: src/test/java/test/project/bookshop/controller/BookControllerTests.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    @Test
+    @WithMockUser(roles = "USER")
+    @DisplayName("Get all books successfully")
+    void getAllBooks_Success() throws Exception {
+        List<BookDto> books = List.of(getBookDto(1L), getBookDto(2L));
+        Page<BookDto> page = new PageImpl<>(books);
+
+        String expectedResponseJson = objectMapper.writeValueAsString(page);
+
+        when(bookService.findAll(any())).thenReturn(page);
```
Comment (liliia-ponomarenko) [inline src/test/java/test/project/bookshop/controller/BookControllerTests.java]: don't use mocks for integration testing. You should add testing data to DB

File: src/test/java/test/project/bookshop/controller/BookControllerTests.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    void getBookById_BookExists_ReturnsExpectedBook() throws Exception {
+        Long bookId = 1L;
+        BookWithoutCategoryIdsDto expectedBook = new BookWithoutCategoryIdsDto();
+        expectedBook.setId(bookId);
+        expectedBook.setTitle("Effective Java");
+        expectedBook.setAuthor("Joshua Bloch");
+        expectedBook.setIsbn("978-0134685991");
+        expectedBook.setPrice(BigDecimal.valueOf(45.00));
+        expectedBook.setDescription("A comprehensive guide to programming in Java");
+        expectedBook.setCoverImage("cover-image-url");
```
Comment (liliia-ponomarenko) [inline src/test/java/test/project/bookshop/controller/BookControllerTests.java]: better move it to the private method or util class and simplify your test case

File: src/test/java/test/project/bookshop/controller/CategoryControllerTests.java [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+import org.springframework.web.context.WebApplicationContext;
+import test.project.bookshop.dto.book.BookDto;
+import test.project.bookshop.dto.category.CategoryDto;
+import test.project.bookshop.dto.category.CategoryRequestDto;
+import test.project.bookshop.service.BookService;
+import test.project.bookshop.service.CategoryService;
+import test.project.bookshop.utils.JwtUtil;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class CategoryControllerTests {
```
Comment (liliia-ponomarenko) [inline src/test/java/test/project/bookshop/controller/CategoryControllerTests.java:38]: the same issues. Let’s  write integration testing correct

File: src/test/java/test/project/bookshop/repository/BookRepositoryTests.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageRequest;
+import test.project.bookshop.model.Book;
+import test.project.bookshop.repository.book.BookRepository;
+
+@DataJpaTest
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+public class BookRepositoryTests {
```
Comment (liliia-ponomarenko) [inline src/test/java/test/project/bookshop/repository/BookRepositoryTests.java:23]: How will you realize your application? Is it okay that it has some default books and categories?
Better add testing data only for tests and remove it after testing.

File: src/test/java/test/project/bookshop/service/BookServiceImplTests.java [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    @Mock
+    private BookMapper bookMapper;
+    @Mock
+    private CategoryService categoryService;
+    @InjectMocks
+    private BookServiceImpl bookService;
+
+    @Test
+    @DisplayName("Save a new book successfully")
+    void save_NewBook_ReturnsSavedBookDto() {
```
Comment (liliia-ponomarenko) [inline src/test/java/test/project/bookshop/service/BookServiceImplTests.java:57]: I recommend you use Given-When-Then construction and increase the readability of your solution.

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix your integration tests and other comments ;)

File: src/test/java/test/project/bookshop/controller/BookControllerTests.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        bookDto.setId(bookId);
+        bookDto.setTitle("Pride and Prejudice");
+        bookDto.setAuthor("Jane Austen");
+        bookDto.setIsbn("978-3-16-148410-0");
+        bookDto.setDescription("A classic novel about love and social standing.");
+        bookDto.setCoverImage("http://example.com/pride_and_prejudice.jpg");
+        bookDto.setPrice(BigDecimal.valueOf(9.99));
+
+        return bookDto;
+    }
```
Comment (Elena-Bruyako) [inline src/test/java/test/project/bookshop/controller/BookControllerTests.java:310]: I suggest to add some not ok test cases,  with invalid data

File: src/test/java/test/project/bookshop/repository/BookRepositoryTests.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Test
+    @DisplayName("Find all books by category ID")
+    void findAllByCategoriesId_CategoryExists_ReturnsBooks() {
+        Long categoryId = 1L;
+        PageRequest pageable = PageRequest.of(0, 10);
+
+        Page<Book> actual = bookRepository.findAllByCategoriesId(categoryId, pageable);
+
+        Assertions.assertEquals(3, actual.getContent().size());
```
Comment (Elena-Bruyako) [inline src/test/java/test/project/bookshop/repository/BookRepositoryTests.java]: let's use static imports (check all places)

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/test/java/test/project/bookshop/repository/BookRepositoryTests.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.test.context.jdbc.Sql;
+import test.project.bookshop.model.Book;
+import test.project.bookshop.repository.book.BookRepository;
+
+@DataJpaTest
+@Sql(scripts = "/db/add-test-books.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
```
Comment (JJJazl) [inline src/test/java/test/project/bookshop/repository/BookRepositoryTests.java:20]:
```suggestion
@Sql(scripts = "/db/add-test-books.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
@Sql(scripts = "/db/clean-data.sql", executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
```

File: src/test/java/test/project/bookshop/service/BookServiceImplTests.java [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+        when(categoryService.findCategoriesByIds(
+                bookRequestDto.getCategoryIds())).thenReturn(categories);
+        when(bookMapper.toBook(bookRequestDto)).thenReturn(firstBook);
+        when(bookRepository.save(firstBook)).thenReturn(firstBook);
+        when(bookMapper.toDto(firstBook)).thenReturn(firstBookDto);
+        BookDto actual = bookService.save(bookRequestDto);
+
+        assertEquals(firstBookDto, actual);
+        verify(categoryService).findCategoriesByIds(bookRequestDto.getCategoryIds());
+        verify(bookRepository).save(firstBook);
```
Comment (JJJazl) [inline src/test/java/test/project/bookshop/service/BookServiceImplTests.java:74]: Let's verify all your mock calls, mappers too. Check everywhere

**Review by JJJazl — APPROVED** [review]
Great job! I am sure you can move on, let's just fix last 2 comments before next tasks)

**Pull request merged**